### PR TITLE
[IMP] Such leave, much attendance

### PR DESCRIPTION
**Pull request by tivisse**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Yenthe666** [comment]
@tivisse are you sure about `poubelle` for naming though? Sounds very French to me 😉

File: addons/hr_payroll/__manifest__.py [hunk 1]
```diff
@@ -22,12 +22,18 @@
         'views/hr_employee_views.xml',
         'data/hr_payroll_sequence.xml',
         'views/hr_payroll_report.xml',
-        'data/hr_payroll_data.xml',
```
Comment (RomainLibert) [inline addons/hr_payroll/__manifest__.py]: unnecessary diff

File: addons/hr_payroll/models/hr_benefit.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+    @api.multi
+    def _duplicate_to_calendar_attendance(self):
+
+        for benefit in self:
+            tz = pytz.timezone(benefit.employee_id.tz)
+            start = pytz.utc.localize(benefit.date_start).astimezone(tz) # in UTC -> timezoned
+            end = pytz.utc.localize(benefit.date_stop).astimezone(tz)
+
+            assert start.date() == end.date()
```
Comment (RomainLibert) [inline addons/hr_payroll/models/hr_benefit.py]: Careful, this assertion will break the loop and rollback what has already been done
Maybe filter out these beforehand

File: addons/hr_payroll/models/hr_leave.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                benefit_type = interval[2].benefit_type_id
+                employee = interval[2].employee_id
+                self.env['hr.benefit'].safe_duplicate_create({
+                    'name': "%s: %s" % (benefit_type.name, employee.name),
+                    'date_start': interval[0],
+                    'date_stop': interval[1],
+                    'benefit_type_id': benefit_type.id,
+                    'employee_id': employee.id,
+                })
+            benefits.unlink()
```
Comment (RomainLibert) [inline addons/hr_payroll/models/hr_leave.py]: Is it right to remove them all even if there was no remaining_benefits?

File: addons/hr_payroll/views/hr_benefit_views.xml [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                <templates>
+                    <t t-name="calendar-box">
+                        <t t-set="color" t-value="widget.getColor(event.color_index)"/>
+                        <div t-att-style="typeof color === 'string' ? ('background-color:'+color)+';' : ''" t-attf-class="#{record.is_highlighted &amp;&amp; record.is_highlighted.value ? 'o_event_hightlight' : ''} #{typeof color === 'number' ? 'o_calendar_color_'+color : ''}">
+                            <field name="name"/>
+                            <div class="fc-time"/>
+                            <div class="o_fields">
+                                <t t-if="record.error">
+                                    <span class="fa fa-exclamation-triangle" aria-label="Conflict" title="error"/>
+                                    <t t-esc="' '"/> <!-- any better way to add a space? ´&nbsp;´ is not recognized :( -->
```
Comment (RomainLibert) [inline addons/hr_payroll/views/hr_benefit_views.xml]: I think you have to use a class to add a margin on the span something like `mr-2`
which is a standard bootstrap 4 class

File: addons/hr_payroll/views/hr_leave_views.xml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<?xml version='1.0' encoding='UTF-8' ?>
+<odoo>
+    <!-- Benefit type on hr leave -->
+    <record id="benefit_type_leave_form_inherit" model="ir.ui.view">
+        <field name="name">benefit.type.leave.form.inherit</field>
+        <field name="model">hr.leave.type</field>
+        <field name="inherit_id" ref="hr_holidays.edit_holiday_status_form"/>
+        <field name="arch" type="xml">
+            <group name="calendar" position="after">
```
Comment (RomainLibert) [inline addons/hr_payroll/views/hr_leave_views.xml]: Please use the xpath tag, as it is more explicit

File: addons/hr_payroll/views/hr_leave_views.xml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+                <group name="benefit" string="Benefits">
+                    <field name="benefit_type_id"/>
+                </group>
+            </group>
+        </field>
+    </record>
+</odoo>
```
Comment (RomainLibert) [inline addons/hr_payroll/views/hr_leave_views.xml]: don't forget to add a line at the end of the file

File: addons/hr_payroll/views/resource_views.xml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <!-- Calendar View -->
+    <record id="payroll_resource_calendar_view_form" model="ir.ui.view">
+        <field name="name">payroll.resource.calendar.view.form</field>
+        <field name="model">resource.calendar</field>
+        <field name="inherit_id" ref="resource.resource_calendar_form"/>
+        <field name="arch" type="xml">
+            <field name="attendance_ids" position="replace">
```
Comment (RomainLibert) [inline addons/hr_payroll/views/resource_views.xml]: Please use xpath tag

File: addons/hr_payroll/views/resource_views.xml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+                <field name="normal_attendance_ids"/>
+            </field>
+            <notebook position="inside">
```
Comment (RomainLibert) [inline addons/hr_payroll/views/resource_views.xml]: Please use xpath tag

File: addons/hr_payroll/wizard/hr_benefit_employee.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+    def generate_benefit(self):
+        if not self.employee_ids:
+            raise UserError(_("You must select employee(s) to generate benefits."))
+        date_start = self.env.context.get('start_benefits')
+        date_stop = self.env.context.get('stop_benefits')
+
+        date_start = fields.Datetime.from_string(date_start)
+        date_stop = fields.Datetime.from_string(date_stop)
+        self.employee_ids.generate_benefit(date_start, date_stop)
```
Comment (RomainLibert) [inline addons/hr_payroll/wizard/hr_benefit_employee.py]: don't forget the empty line at the end of the file

File: addons/hr_payroll/wizard/hr_payroll_payslips_by_employees.py [hunk 10]
```diff
@@ -3,18 +3,22 @@
...
+from datetime import datetime
 
 
 class HrPayslipEmployees(models.TransientModel):
     _name = 'hr.payslip.employees'
     _description = 'Generate payslips for all selected employees'
 
     employee_ids = fields.Many2many('hr.employee', 'hr_employee_group_rel', 'payslip_id', 'employee_id', 'Employees',
-                                    domain=lambda self: self._get_available_contracts_domain())
+                                    domain=lambda self: self._get_available_contracts_domain(),
```
Comment (RomainLibert) [inline addons/hr_payroll/wizard/hr_payroll_payslips_by_employees.py]: put those two method above field definition as per guidelines

**Comment by tivisse** [comment]
Done at https://github.com/odoo/odoo/commit/bdc9688d289bebedad068a889d4ce79b19fbe915

File: addons/hr_payroll/static/src/js/benefit.js [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                self._renderBenefitButtons();
+            });
+        },
+
+        _renderGenerateButton: function(date_from, date_to, employee_ids, secondary) {
+            var self = this;
+            var primary = !secondary ? 'btn-primary' : 'btn-secondary';
+            var txt = _t("Generate Benefits");
+            this.$buttons.find('.o_calendar_button_month').after(
+                $('<button class="btn ' + primary + ' btn-benefit" type="button">'+ txt +'</button>')
```
Comment (odony) [inline addons/hr_payroll/static/src/js/benefit.js]: This is an **unsafe** way of inserting text in a dom element. It may not be dangerous here but it's an anti-pattern that you must avoid at all costs. Use a proper QWeb template whenever possible, and when it's really overkill use the proper JQuery primitives such as addClass() and text() to avoid stupid XSS injections.

e.g.:
```var $button = $('<button class="btn" type="button">').addClass(primary).text(txt)```

File: addons/hr_payroll/static/src/js/benefit.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+                    };
+                    self.do_action({
+                        type: 'ir.actions.act_window',
+                        name: txt,
+                        res_model: 'hr.benefit.employees',
+                        view_type: 'form',
+                        views: [[false,'form']],
+                        target: 'new',
+                        context: {
+                            'start_benefits': date_from.format(date_fmt),
```
Comment (odony) [inline addons/hr_payroll/static/src/js/benefit.js]: Don't roll your own formatting every time you need to format a value, use the framework, e.g. `datetime_to_str()`

File: addons/hr_payroll/static/src/js/benefit.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                                default_date_start: firstDay.format(date_fmt),
+                                default_date_end: lastDay.format(date_fmt),
+                            },
+                        });
+                    })
+                );
+            }
+            else if (!is_validated) { // Validate button
+                this.$buttons.find('.o_calendar_button_month').after(
+                    $('<button class="btn btn-primary btn-benefit" type="button">'+ _t('Validate Benefits') +'</button>')
```
Comment (odony) [inline addons/hr_payroll/static/src/js/benefit.js]: same thing as above. don't handcraft DOM source code, use the proper (and safe) JQuery API!

File: addons/hr_payroll/tests/test_benefit.py [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+from odoo.addons.hr_payroll.tests.common import TestPayslipBase
+
+
+@tagged('benefit')
+class TestBenefit(TestPayslipBase):
+
+    def setUp(self):
+        super(TestBenefit, self).setUp()
+        self.tz = pytz.timezone(self.richard_emp.tz)
+        self.start = datetime.strptime('2015-11-01 01:00:00', '%Y-%m-%d %H:%M:%S').replace(tzinfo=self.tz)
```
Comment (odony) [inline addons/hr_payroll/tests/test_benefit.py]: You should never need to use strptime(), the framework has all the bells and whistles to do that for you! E.g. `fields.Datetime.to_datetime()`. Please get rid of them all.

File: addons/hr_payroll/tests/test_benefit.py [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+        benefits = benefit.split_by_day()
+        self.assertEqual(len(benefits), 3, "Benefit should be split in three")
+        self.assertEqual(pytz.utc.localize(benefits[0].date_start), self.tz.localize(datetime.strptime('2015-11-01 09:00:00', '%Y-%m-%d %H:%M:%S')))
+        self.assertEqual(pytz.utc.localize(benefits[0].date_stop), self.tz.localize(datetime.strptime('2015-11-01 23:59:59', '%Y-%m-%d %H:%M:%S')))
+
+        self.assertEqual(pytz.utc.localize(benefits[1].date_start), self.tz.localize(datetime.strptime('2015-11-02 00:00:00', '%Y-%m-%d %H:%M:%S')))
+        self.assertEqual(pytz.utc.localize(benefits[1].date_stop), self.tz.localize(datetime.strptime('2015-11-02 23:59:59', '%Y-%m-%d %H:%M:%S')))
+
+        self.assertEqual(pytz.utc.localize(benefits[2].date_start), self.tz.localize(datetime.strptime('2015-11-03 00:00:00', '%Y-%m-%d %H:%M:%S')))
+        self.assertEqual(pytz.utc.localize(benefits[2].date_stop), self.tz.localize(datetime.strptime('2015-11-03 18:00:00', '%Y-%m-%d %H:%M:%S')))
```
Comment (odony) [inline addons/hr_payroll/tests/test_benefit.py]: please make helper functions for all that redundant localize stuff, it's ugly and unreadable

File: addons/hr_payroll/views/hr_benefit_views.xml [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+                        <div t-att-style="typeof color === 'string' ? ('background-color:'+color)+';' : ''" t-attf-class="#{record.is_highlighted &amp;&amp; record.is_highlighted.value ? 'o_event_hightlight' : ''} #{typeof color === 'number' ? 'o_calendar_color_'+color : ''}">
+                            <field name="name"/>
+                            <div class="fc-time"/>
+                            <div class="o_fields">
+                                <t t-if="record.display_warning">
+                                    <span class="fa fa-exclamation-triangle mr-2" aria-label="Conflict" title="error"/>
+                                </t>
+                                <t t-if="record.state == 'validated'">
+                                    <span class="fa fa-check-circle-o" aria-label="Validated" title="validated"/>
+                                    <t t-esc="' '"/>
```
Comment (odony) [inline addons/hr_payroll/views/hr_benefit_views.xml]: WTF?!

File: addons/hr_payroll/views/hr_benefit_views.xml [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+        <field name="name">hr.benefit.form</field>
+        <field name="model">hr.benefit</field>
+        <field name="arch" type="xml">
+            <form string="Benefit" >
+                <header>
+                    <field name="display_warning" invisible="1"/>
+                    <field name="leave_id" invisible="1"/>
+                    <button name="action_cancel" states="confirmed" string="Cancel" type="object"/>
+                    <button name="action_draft" states="cancelled" string="Reset to Draft" type="object"/>
+                    <button string="Go To Conflicting Leave" name="action_leave" type="object" attrs="{'invisible': ['|', '!',('display_warning', '=', True), '!', ('leave_id', '!=', False)]}"/>
```
Comment (odony) [inline addons/hr_payroll/views/hr_benefit_views.xml]: `not leave_id != False` => `leave_id = False` ??

File: addons/hr_payroll/wizard/hr_payroll_payslips_by_employees.py [hunk 18]
```diff
@@ -3,17 +3,22 @@
...
 class HrPayslipEmployees(models.TransientModel):
     _name = 'hr.payslip.employees'
     _description = 'Generate payslips for all selected employees'
 
-    employee_ids = fields.Many2many('hr.employee', 'hr_employee_group_rel', 'payslip_id', 'employee_id', 'Employees',
-                                    domain=lambda self: self._get_available_contracts_domain())
-
     def _get_available_contracts_domain(self):
-        return [('contract_ids.state', 'in', ('open', 'pending'))]
+        return [('contract_ids.state', 'in', ('open', 'pending')), ('company_id', '=', self.env.user.company_id.id)]
```
Comment (odony) [inline addons/hr_payroll/wizard/hr_payroll_payslips_by_employees.py]: once again, shouldn't the per-company filtering be done by the default record rules? why hardcode the rules, preventing any customisation by the admin?

File: addons/hr_payroll/wizard/hr_payroll_payslips_by_employees.py [hunk 19]
```diff
@@ -23,8 +28,20 @@ def compute_sheet(self):
         active_id = self.env.context.get('active_id')
         if active_id:
             [run_data] = self.env['hr.payslip.run'].browse(active_id).read(['date_start', 'date_end', 'credit_note'])
-        from_date = run_data.get('date_start')
-        to_date = run_data.get('date_end')
+            from_date = run_data.get('date_start')
+            to_date = run_data.get('date_end')
+        else:
+            from_date = datetime.strptime(self.env.context.get('default_date_start'), '%Y-%m-%d')
```
Comment (odony) [inline addons/hr_payroll/wizard/hr_payroll_payslips_by_employees.py]: use Datetime.to_datetime() everywhere you used strptime please.

File: addons/resource/models/resource.py [hunk 20]
```diff
@@ -197,21 +197,28 @@ def _onchange_hours_per_day(self):
...
+    def _attendance_intervals(self, start_dt, end_dt, resource=None, domain=None):
         """ Return the attendance intervals in the given datetime range.
             The returned intervals are expressed in the resource's timezone.
         """
         assert start_dt.tzinfo and end_dt.tzinfo
         combine = datetime.combine
 
+        resource_ids = [resource.id, False] if resource else [False]
+        domain = domain if domain is not None else []
+        domain = domain + [
```
Comment (odony) [inline addons/resource/models/resource.py]: concatenating arbitrary domains is done with expression.AND not with addition

File: addons/resource/models/resource_mixin.py [hunk 21]
```diff
@@ -53,6 +53,47 @@ def copy_data(self, default=None):
...
+            for day in day_hours
+        )
+        return {
+            'days': days,
+            'hours': sum(day_hours.values()),
+        }
+
+    def _get_day_total(self, from_datetime, to_datetime, calendar, resource):
+        """
+        @return dict with hours of attendance in each day between `from_datetime` and `to_datetime`
```
Comment (odony) [inline addons/resource/models/resource_mixin.py]: user proper docstring format if you write docstrings

File: addons/resource/models/resource_mixin.py [hunk 22]
```diff
@@ -53,6 +53,47 @@ def copy_data(self, default=None):
...
+        # in order to compute the total hours on the first and last days
+        from_full = from_datetime - timedelta(days=1)
+        to_full = to_datetime + timedelta(days=1)
+        intervals = calendar._attendance_intervals(from_full, to_full, resource)
+        day_total = defaultdict(float)
+        for start, stop, meta in intervals:
+            day_total[start.date()] += (stop - start).total_seconds() / 3600
+        return day_total
+
+    def _timezone_datetime(self, time):
```
Comment (odony) [inline addons/resource/models/resource_mixin.py]: why is this a method at all? it should be a helper function

File: addons/resource/models/resource_mixin.py [hunk 23]
```diff
@@ -68,38 +109,18 @@ def get_work_days_data(self, from_datetime, to_datetime, compute_leaves=True, ca
...
-            float_utils.round(ROUNDING_FACTOR * day_hours[day] / day_total[day]) / ROUNDING_FACTOR
-            for day in day_hours
-        )
-        return {
-            'days': days,
-            'hours': sum(day_hours.values()),
-        }
+        return self._get_days_data(intervals, day_total)
 
     def get_leave_days_data(self, from_datetime, to_datetime, calendar=None, domain=None):
```
Comment (odony) [inline addons/resource/models/resource_mixin.py]: good opportunity to make this method private, it should not be public, right?

File: addons/resource/models/resource_mixin.py [hunk 24]
```diff
@@ -53,6 +53,47 @@ def copy_data(self, default=None):
...
+            day_total[start.date()] += (stop - start).total_seconds() / 3600
+        return day_total
+
+    def _timezone_datetime(self, time):
+        if not time.tzinfo:
+            time = time.replace(tzinfo=utc)
+        return time
+
+
     def get_work_days_data(self, from_datetime, to_datetime, compute_leaves=True, calendar=None, domain=None):
```
Comment (odony) [inline addons/resource/models/resource_mixin.py]: good opportunity to make this method private, it should not be public, right?

File: addons/hr_payroll/__manifest__.py [hunk 25]
```diff
@@ -28,6 +28,12 @@
         'views/report_contributionregister_templates.xml',
         'views/report_payslip_templates.xml',
         'views/report_payslipdetails_templates.xml',
+        'views/hr_benefit_views.xml',
+        'views/hr_benefit_views.xml',
```
Comment (mreficent) [inline addons/hr_payroll/__manifest__.py]: duplicated

File: addons/hr_payroll/models/hr_benefit.py [hunk 26]
```diff
@@ -0,0 +1,296 @@
...
+        """
+        try:
+            with mute_logger('odoo.sql_db'), self.env.cr.savepoint():
+                # Any database call will be run inside a postgresql savepoint which is
+                # rolledback if an exception occurs allowing to make subsequent calls in the transaction.
+                # (Otherwise InternalError is raised by any subsequent database operations)
+                return self.create(vals)
+        except IntegrityError as err:
+            if 'hr_benefit__unique' not in err.pgerror:
+                raise err
```
[the remark this answers is not in the document]
Comment (tivisse) [reply addons/hr_payroll/models/hr_benefit.py]: I fear that testing for conflicts will be quite heavy. Why not doing something like in the translations ? A massive SQL insert with a "ON CONFLICT DO NOTHING" clause ?

[the remark this answers is not in the document]
Comment (tivisse) [reply addons/hr_payroll/models/hr_benefit.py]: What about this ? https://github.com/odoo/odoo/pull/31091/commits/8e4d7bf8b9dacc3b817db2ab6b783ebb524bb4f1

File: addons/hr_payroll/models/hr_benefit.py [hunk 27]
```diff
@@ -0,0 +1,296 @@
...
+from odoo.addons.resource.models.resource import Intervals
+
+class HrBenefit(models.Model):
+    _name = 'hr.benefit'
+    _description = 'hr.benefit'
+
+    name = fields.Char(required=True)
+    active = fields.Boolean(default=True)
+    employee_id = fields.Many2one('hr.employee', required=True,
+        domain=lambda self: [('contract_ids.state', 'in', ('open', 'pending')), ('company_id', '=', self.env.user.company_id.id)])
```
[the remark this answers is not in the document]
Comment (tivisse) [reply addons/hr_payroll/models/hr_benefit.py]: There are no multi company record rule on employees. But we can removed it.

[the remark this answers is not in the document]
Comment (odony) [reply addons/hr_payroll/models/hr_benefit.py]: Wouldn't it make sense to have this record rule? Was it removed on purpose?
Additionally, I assume that most "hr.benefit" records will be created automatically, and HR Officers will mostly review/edit them, right? If that's correct, then selecting a different employee won't be a common use case, and the domain is a bit useless.

[the remark this answers is not in the document]
Comment (Musvol) [reply addons/hr_payroll/models/hr_benefit.py]: You are right, the main use case for HR is reviewing `hr.benefit`.
On the other hand, if one benefit is manually created with an employee from another company, it will affect payslips of the other company which we don't want. We should add multi company record rules either on `hr.employee` or at least on `hr.benefit` to avoid that.

[the remark this answers is not in the document]
Comment (tivisse) [reply addons/hr_payroll/models/hr_benefit.py]: It was removed on purpose: "Someone" would like to see all the employees to make stats. 

I've already created several feedbacks to reintroduce them, always refused.

We should add an ir.rule on benefits (Mostly for privacy and to avoid generating payslips for indian people, ...).

File: addons/hr_payroll/views/hr_benefit_views.xml (hunk 16 above)
Comment (Musvol) [reply addons/hr_payroll/views/hr_benefit_views.xml]: I tried several other things to add a space such as `&nbsp;`.  None worked in the context of the calendar template.
I'll try to find a less hackinsh way and keep a nice layout.

File: addons/hr_payroll/models/hr_benefit.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (odony) [reply addons/hr_payroll/models/hr_benefit.py]: > It was removed on purpose: "Someone" would like to see all the employees to make stats.

We have general purpose multi-company rules everywhere, this one should be added for consistency. Customising record rules for specific companies is something that is done by the admin, on a case by case basis.
@Musvol If there is a real risk of messing up data, the sanity check should be done in the business code before we perform the actual action, rather than hoping all is well because there was a domain on the employee_id (employees could theoretically be reassigned to other companies, etc.).

File: addons/hr_payroll/models/hr_leave.py [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+                    'date_stop': benefit_stop,
+                    'benefit_type_id': benefit_type.id,
+                    'employee_id': employee.id,
+                })
+            benefits.unlink()
+
+    @api.multi
+    def action_validate(self):
+        super(HrLeave, self).action_validate()
+        self._cancel_benefit_conflict()
```
[the remark this answers is not in the document]
Comment (Musvol) [reply addons/hr_payroll/models/hr_leave.py]: Actually, it does not add much protection to forbid the access of benefits for Holidays Officer/Manager/Team Leader:

benefits are only an interface to create `resource.calendar.[leaves | attendance]` when they are validated.
If we grant them access, the only additional info we would give are employees' work schedules, which is not private info, is it?
If a user from Holidays group tries to validate benefits, it would be ok for  `resource.calendar.leaves` but throw an AccessError because they lack access for `resource.calendar.attendance`.

To solve the issue above, I would give access rights to benefits for Holidays groups.
What do you think? Would you prefer not and use `sudo()` for holiday users in `action_validate()` and `action_refuse()` ?

[the remark this answers is not in the document]
Comment (odony) [reply addons/hr_payroll/models/hr_leave.py]: We should try to follow the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege). If a role does not *need* access to a table/record, then they should not have it, especially not _write_ access.
It seems risky to let "team leaders" mess up with actual hr.benefit entries, even if only by mistake (reminder: the access won't be restricted to just `action_validate()`, they will have full CRUD). These benefits records are meant to be used for payroll purposes, ultimately generating payslips, so they must be carefully controlled and verified by Payroll staff. In some companies it will be unthinkable to give this power to team leaders.
So yes, it seems you should probably sudo() just for those calls.

File: addons/hr_payroll/wizard/hr_payroll_payslips_by_employees.py (hunk 18 above)
Comment (tivisse) [reply addons/hr_payroll/wizard/hr_payroll_payslips_by_employees.py]: The ir.rule comes in 12.2 sadly.
